### DISCUSSION TITLE
Instant Crash without Play Services

### DISCUSSION THREAD

**Discussion by moberer** [NightscoutFoundation/xDrip#2578] [General]
(Created this as a discussion, not an issue, since issue creation dialog states to do so.)

**Environment:**
- xDrip version 20221116-344deb4
- Pixel 6a with GrapheneOS, without Google Play or MicroG

**Action to reproduce:**
- Open app on a user profile without Google Play Services.

**Result:**
- App flashes up and then crashes in under half a second.

**Expected Result:**
- App should open and be usable.

**Workaround:**
- The app works on a profile with Google Play installed.

**Notes:**
I saw some old issues and discussion about Google Play / MicroG being required.
However, from what I could take away, it is only ever used for notifications for Followers.
If this is really the case, I think the rest of the functionality should work without Play Services.

In any case: The app should not hard-crash in a case like this.

**Further Notes:** 
I also found some discussion about "AppMeasurementService" on StackOverflow.
It seems that Android Studio sometimes adds this on its own to apps.
In this case, fixing this should be as easy as removing the auto-added stuff.
See e.g. https://stackoverflow.com/questions/33366846/what-is-appmeasurementservice-and-why-does-android-studio-add-this-service-to-my for some info on this.

**Backtrace:**
```
type: crash
osVersion: google/bluejay/bluejay:13/TQ1A.221205.011/2022122000:user/release-keys
package: com.eveningoutpost.dexdrip:1603091400
process: com.eveningoutpost.dexdrip
processUptime: 475 + 189 ms

java.lang.RuntimeException: Unable to create service com.google.android.gms.measurement.AppMeasurementService: java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.measurement.AppMeasurementService" on path: DexPathList[[zip file "/data/app/~~eysawNWmQ4hJYEcbeJV8Dw==/com.eveningoutpost.dexdrip-R3zFO1xYfLPEK29EI55vNw==/base.apk"],nativeLibraryDirectories=[/data/app/~~eysawNWmQ4hJYEcbeJV8Dw==/com.eveningoutpost.dexdrip-R3zFO1xYfLPEK29EI55vNw==/lib/arm64, /system/lib64, /system_ext/lib64]]
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4504)
	at android.app.ActivityThread.-$$Nest$mhandleCreateService(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2162)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7878)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ExecInit.main(ExecInit.java:49)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:355)
Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.measurement.AppMeasurementService" on path: DexPathList[[zip file "/data/app/~~eysawNWmQ4hJYEcbeJV8Dw==/com.eveningoutpost.dexdrip-R3zFO1xYfLPEK29EI55vNw==/base.apk"],nativeLibraryDirectories=[/data/app/~~eysawNWmQ4hJYEcbeJV8Dw==/com.eveningoutpost.dexdrip-R3zFO1xYfLPEK29EI55vNw==/lib/arm64, /system/lib64, /system_ext/lib64]]
	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:259)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
	at android.app.AppComponentFactory.instantiateService(AppComponentFactory.java:129)
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4472)
	... 11 more

```

Thanks for your time and effort!

**Comment by Navid200** [comment]
Thanks for reporting.

Pixel 6a comes with Google play service installed already, doesn't it?
If yes, may I ask how you uninstalled it?

My understanding is that xDrip should come up OK without Google Play service.  However, there will be constant notifications asking you to install it.  And, there is a setting in xDrip you can enable to stop getting those notifications.
But, if it crashes, obviously, there will be no way to change any settings.

I have a few phones and wish to test this.  However, I don't have a Pixel phone.  Nevertheless, I would still like to attempt recreating the problem.
Would you please tell me how I can uninstall Google Play service?   Is this something I can do without having to root the phone?

**Reply by moberer** [reply]
Indeed, from the factory it does have Google Play Services. I do not think you can uninstall them on the factory image.

I've flashed a custom ROM (GrapheneOS), which is based on AOSP and does not come with any Play Services.


I think the best way to get a Play-Less hardware device is flashing it with a custom ROM. (Lineage, /e/, etc.) - Though I understand if you don't necessarily want to do that.

However, there is always the possibility of using an Android emulator: E.g. like https://stackoverflow.com/questions/33100132/how-to-set-up-android-emulator-without-google-play-services

**Reply by moberer** [reply]
Also, if I can help by testing something on my device, feel free to ask me.

**Reply by Navid200** [reply]
It may help if you edit your first post here and clarify exactly which version(s) of xDrip this has happened with.

I know that one can just disable Google Play services on their phone.  And I know that in that case, xDrip runs fine on the phone except that there will be no xDrip sync capability.
I also know that some Huawei phones are sold without Google Play service on them.  Users have reported that xDrip runs fine on them.  Of course, again, excluding xDrip sync.

Is it possible that something else on the image you are using is the cause of this?
Have you tried any other image that excludes Google Play?  Do they all crash?

**Reply by moberer** [reply]
Added the version information. - It's the latest release, downloaded today.

> Is it possible that something else on the image you are using is the cause of this?

Yes, there are three major things I can think of right now:

GrapheneOS uses a different memory allocator (malloc) implementation, which is designed to be more resilient against exploits. - This means lingering memory errors (use-after-free, etc.) are more likely to cause a segfault.
I do not think this is the issue here, because
- the backtrace does not show a segfault
- the application is written in Java, which makes such errors unlikely

GrapheneOS enables Play Services to be installed only for a single user by some sandboxing tricks. I also do not think that this is the problem, because:
- this behavior is only enabled if Play Services /are/ installed and
- if Play Services /are/ installed on a given user profile, the app works flawlessly.

Furthermore, on a Play-Service-less user profile, there is /no/ alternative installed: That is there is also /no/ microG.


> Have you tried any other image that excludes Google Play? Do they all crash?

Unfortunately, I don't have a secondary device to do that on. - And I don't want to reflash my primary phone.

The things I have tried:
- xDrip running on Pixel 6a in a user profile with /no/ Play Services installed: crashes
- xDrip running on Pixel 6a in a user profile /with/ Play Services installed*: connects to sensor, works.

**Reply by Navid200** [reply]
Can you use a virtual machine to test other images?

Can you tell me if I can find an image that fails for the Android Studio virtual device?
https://developer.android.com/studio/run/managing-avds
That's what I can use for running tests myself.

**Reply by Navid200** [reply]
I see that you are also providing advice in PRs.
If you can fix this and open a PR, please go ahead.
I have a lot on my plate already.

**Reply by moberer** [reply]
Well, I can try to and see how far I get.
My programming experience is mainly in different areas, not Java/Android.

**Reply by Navid200** [reply]
Would you please install this version of xDrip and see if that crashes too?
https://github.com/NightscoutFoundation/xDrip/releases/tag/2020.11.27

**Reply by moberer** [reply]
It seems to work, yes! - At least it doesn't crash at startup.

I'll report back in a few more minutes if the rest of the in-app functionality works.

**Reply by moberer** [reply]
Have just paired a sensor and everything works as expected!

Thanks a whole lot!

**Reply by Navid200** [reply]
How about this?  Does this one also start without a crash?
https://github.com/NightscoutFoundation/xDrip/releases/tag/2021.07.31

**Reply by moberer** [reply]
Yes, but I get a small "Google Play not installed" notification in the middle bottom area of the screen, which wasn't there with the other build. (Though I understand that is expected)

**Reply by Navid200** [reply]
Would you please try this one?  Does it start or does it crash? 
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.04.20

**Reply by moberer** [reply]
It starts normally.

**Reply by Navid200** [reply]
Thanks for your patience.
Would you please try this one?  Does it also start without a crash?
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.08.18

**Reply by moberer** [reply]
No problem!

It starts normally.

**Reply by Navid200** [reply]
Would you please try this one?
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.10.25

**Reply by moberer** [reply]
It crashes.

**Reply by Navid200** [reply]
OK.
Does this one crash?
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.10.15

**Reply by moberer** [reply]
Crashes.

**Reply by Navid200** [reply]
I didn't expect that.
We are almost there.  Sorry about all of this and thanks for your patience.
Would you please try this?
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.21

**Reply by moberer** [reply]
No problem at all. - At least this way I can save you a lot of time narrowing it down.

Still crashes.

**Reply by Navid200** [reply]
Thanks.  
Would you please try this now?
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.08.31

**Reply by moberer** [reply]
Does a bit of a weird white flash when starting, but works otherwise. - I think that is unrelated.

Do you use bisect, or are you just manually narrowing it down?

**Reply by Navid200** [reply]
I was doing binary search until October 25, when I had an idea.  Which was wrong.  After that continuing binary search.  
At this point we have August 31 working and September 21 crashing.  And the only releases between those two dates are the following:
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.01
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.02
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.06
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.08b

Would you mind trying those and telling me which ones crash?

**Reply by moberer** [reply]
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.01 works
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.02 works
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.06 works
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.08b crashes

**Reply by moberer** [reply]
https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.09.08 works (without b)

**Reply by Navid200** [reply]
Thanks

If you are certain about that (2022.09.08 working and 2022.09.08b crashing), please go ahead and either convert this discussion to an issue, or open a new issue.
If you open a new issue (not converting this discussion), please include a link to this discussion in the first post of the issue.

Thanks for reporting

**Reply by moberer** [reply]
Yes, will do. Thanks a lot for your time and effort!
I'll also try and build a reproducible emulator environment before I open as an issue.

**Comment by jamorham** [comment]
Please confirm if this problem is reproducible in the emulator? As far as I know appmeasurement is turned off as xDrip doesn't use it, but its possible something has gone wrong when we updated google play service versions. The stack-trace doesn't show anything that is actually part of xDrip so it must be being initiated somewhere else.

**Reply by moberer** [reply]
I was downloading all the tools, but in the meantime, ~~it has been resolved~~ here: https://github.com/NightscoutFoundation/xDrip/discussions/2578#discussioncomment-4488583

**Reply by Navid200** [reply]
That was not meant to be a resolution.   If you don't mind, please work with me.  I would like to figure out exactly which version causes the problem.
